### PR TITLE
Offer base remediation

### LINKED ISSUE #505: [OBE-01M] Inexistent Integration of Unlimited Offers
## OBE-01M: Inexistent Integration of Unlimited Offers

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) |  | OfferBase.sol:L320 |

### Description:

The Boson Protocol system supports a special flag for the `quantityAvailable` (`type(uint256).max`) meant to indicate that it should not be reduced when new vouchers are minted for it, however, this flag is not supported by the premint system.

### Impact:

The current implementation of premints deviates from the original behaviour of standard voucher mints which is undesirable.

### Example:

```solidity
function reserveRangeInternal(uint256 _offerId, uint256 _length) internal {
    // Get offer, make sure the caller is the operator
    Offer storage offer = getValidOffer(_offerId);

    // Prevent reservation of an empty range
    require(_length > 0, INVALID_RANGE_LENGTH);

    // Cannot reserve more than it's available
    require(offer.quantityAvailable >= _length, INVALID_RANGE_LENGTH);

    // Prevent reservation of too large range, since it affects exchangeId
    require(_length < (1 << 128), INVALID_RANGE_LENGTH);

    // Get starting token id
    ProtocolLib.ProtocolCounters storage pc = protocolCounters();
    uint256 _startId = pc.nextExchangeId;

    // Call reserveRange on voucher
    IBosonVoucher bosonVoucher = IBosonVoucher(protocolLookups().cloneAddress[offer.sellerId]);
    bosonVoucher.reserveRange(_offerId, _startId, _length);

    // increase exchangeIds
    pc.nextExchangeId = _startId + _length;

    // decrease quantity available
    offer.quantityAvailable -= _length;

    // Notify external observers
    emit RangeReserved(_offerId, offer.sellerId, _startId, _startId + _length - 1, msgSender());
}
```

### Recommendation:

We advise it to be done so, ensuring the premint system conforms to the specifications of the regular mint system for vouchers.

### LINKED ISSUE #506: [OBE-02M] Unsatisfied Function Pre-Conditions
## OBE-02M: Unsatisfied Function Pre-Conditions

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) |  | OfferBase.sol:L282, [L283](https://github.com/bosonprotocol/boson-protocol-contracts/blob/38c7ba384e55486aee4ea76bb80b4f29d1a86175/contracts/protocol/bases/OfferBase.sol#L283) |

### Description:

The `reserveRangeInternal` function is meant to fail if the offers or exchanges region of the protocol is paused, however, no such validation is performed at the function level.

### Impact:

It is currently possible to invoke `reserveRangeInternal` via multiple methods bypassing its security assumption. In detail, the exchange region assumption is invalidated by the following functions

### Example:

```solidity
/**
 * @notice Reserves a range of vouchers to be associated with an offer
 *
 * Reverts if:
 * - The offers region of protocol is paused
 * - The exchanges region of protocol is paused
 * - Offer does not exist
 * - Offer already voided
 * - Caller is not the seller
 * - Range length is zero
 * - Range length is greater than quantity available
 * - Range length is greater than maximum allowed range length
 * - Call to BosonVoucher.reserveRange() reverts
 *
 * @param _offerId - the id of the offer
 * @param _length - the length of the range
 */
function reserveRangeInternal(uint256 _offerId, uint256 _length) internal {
    // Get offer, make sure the caller is the operator
    Offer storage offer = getValidOffer(_offerId);

    // Prevent reservation of an empty range
    require(_length > 0, INVALID_RANGE_LENGTH);

    // Cannot reserve more than it's available
    require(offer.quantityAvailable >= _length, INVALID_RANGE_LENGTH);

    // Prevent reservation of too large range, since it affects exchangeId
    require(_length < (1 << 128), INVALID_RANGE_LENGTH);

    // Get starting token id
    ProtocolLib.ProtocolCounters storage pc = protocolCounters();
    uint256 _startId = pc.nextExchangeId;

    // Call reserveRange on voucher
    IBosonVoucher bosonVoucher = IBosonVoucher(protocolLookups().cloneAddress[offer.sellerId]);
    bosonVoucher.reserveRange(_offerId, _startId, _length);

    // increase exchangeIds
    pc.nextExchangeId = _startId + _length;

    // decrease quantity available
    offer.quantityAvailable -= _length;

    // Notify external observers
    emit RangeReserved(_offerId, offer.sellerId, _startId, _startId + _length - 1, msgSender());
}
```

### Recommendation:

We advise the function to validate those two system modules being active as none of its internal or external call chains (`getValidOffer`, `protocolCounters`, `protocolLookups`, `reserveRange`) do so in the current implementation. While both regions are validated in the `reserveRange` function of `OfferHandlerFacet`, the same validation is not present in other invocations of the system such as within `OrchestrationHandlerFacet1`.

### PR DESCRIPTION
**Pull request by zajck**
Fixes #505 
Fixes #506

### CODE REVIEW

**Review by mischat — APPROVED** [review]
This looks good to me, and it is a great catch!

File: contracts/protocol/bases/OfferBase.sol [hunk 1]
```diff
@@ -292,7 +292,7 @@ contract OfferBase is ProtocolBase, IBosonOfferEvents {
      * @param _offerId - the id of the offer
      * @param _length - the length of the range
      */
-    function reserveRangeInternal(uint256 _offerId, uint256 _length) internal {
+    function reserveRangeInternal(uint256 _offerId, uint256 _length) internal offersNotPaused exchangesNotPaused {
```
Comment (mischat) [inline contracts/protocol/bases/OfferBase.sol:295]: why did you choose to remove the `nonReentrant` from here ?

File: contracts/protocol/facets/OfferHandlerFacet.sol [hunk 2]
```diff
@@ -141,13 +141,7 @@ contract OfferHandlerFacet is IBosonOfferHandler, OfferBase {
      * @param _offerId - the id of the offer
      * @param _length - the length of the range
      */
-    function reserveRange(uint256 _offerId, uint256 _length)
-        external
-        override
-        nonReentrant
```
Comment (mischat) [inline contracts/protocol/facets/OfferHandlerFacet.sol:147]: My previous comment is related to this `nonReenterant` modifier

**Review by mischat — COMMENTED** [review]
I have left a comment @zajck

File: contracts/protocol/facets/OfferHandlerFacet.sol (hunk 2 above)
Comment (zajck) [reply contracts/protocol/facets/OfferHandlerFacet.sol:147]: `nonReentrant` is still here. Prettier just put it into the same line with function name and other stuff.

**Review by albertfolch-redeemeum — APPROVED** [review]

**Pull request merged**